app/Http/Controllers/ConversionMetricController.php

```diff
@@ -0,0 +1,79 @@
+<?php
+
+namespace App\Http\Controllers;
+
+use App\Wiki;
+use Carbon\CarbonImmutable;
+use Illuminate\Http\Response;
+use Illuminate\Http\Request;
+
+class ConversionMetricController extends Controller
+{
+
+    private string $fileName="conversion_metric_for_all_wikis.csv";
+
+    /**
+     * Produce a downloadable csv file with conversion metrics for all wikis.
+     */
+    public function index(Request $request)
+    {
+        $allWikis = Wiki::all();
+        $current_date = CarbonImmutable::now();
+        $output = [];
+        
+
+        foreach ($allWikis as $wiki) {
+            $lifecycleEvents = $wiki->wikiLifecycleEvents()->first();
+            $wikiLastEditedTime = null;
+            $wikiFirstEditedTime = null;
+            if ($lifecycleEvents !== [] && $lifecycleEvents->count() > 0) {
+                if ($lifecycleEvents['last_edited']) {
+                    $wikiLastEditedTime = CarbonImmutable::parse($lifecycleEvents['last_edited']);    
+                }
+                if ($lifecycleEvents['first_edited']) {
+                    $wikiFirstEditedTime = CarbonImmutable::parse($lifecycleEvents['first_edited']);
+                }
+            }
+            $time_before_wiki_abandoned_days = null;
+            $time_to_engage_days = null;
+
+            if (!is_null($wikiLastEditedTime) && ($current_date->diffInDays($wikiLastEditedTime) >= 90)) {
+                $time_before_wiki_abandoned_days = $wikiLastEditedTime->diffInDays($wiki->created_at);
+            }
+            if ($wikiFirstEditedTime !== null) {
+                echo "FirstEdited:" . $wikiFirstEditedTime . "CreatedAt:" .$wiki->created_at . "\n";
+                $time_to_engage_days = $wikiFirstEditedTime->diffInDays($wiki->created_at);
+            }
+            $wiki_number_of_editors = $wiki->wikiSiteStats()->first()['activeusers'] ?? null;
+
+            $output[] = [
+                'domain' => $wiki->domain,
+                'time_to_engage_days' => $time_to_engage_days,
+                'time_before_wiki_abandoned_days' => $time_before_wiki_abandoned_days,
+                'number_of_active_editors' => $wiki_number_of_editors
+            ];
+
+        }
+
+        if ( $request->wantsJson() ) {
+            return response()->json( $output );
+        }
+
+        return $this->returnCsv($output);
+
+    }
+
+    private function returnCsv( $output ) {
+        ob_start();
+		$handle = fopen('php://output', 'r+');
+        fputcsv($handle, ['domain_name', 'time_to_engage_days', 'time_before_wiki_abandoned_days', 'number_of_active_editors']);
+        foreach ($output as $wikiMetrics) {
+            fputcsv($handle, array_values($wikiMetrics));
+        }
+        $csv = $csv = ob_get_clean();
+        return response($csv, 200, [
+			'Content-Type' => 'text/csv',
+			'Content-Disposition' => 'attachment;filename='.$this->fileName
```

Comment:
suggestion: personally I think it's extremely useful to have some timestamp along a CSV file, we could add something like
```php
$timestamp = date('Y-m-d_H-i-s_', request()->server('REQUEST_TIME'));
```

Reply:
Nice idea!